otx/algorithms/common/adapters/torch/utils/bs_search_algo.py

```diff
@@ -12,66 +12,199 @@
-            break
+    @staticmethod
+    def _get_even_center_val(val1: int, val2: int) -> int:
+        ret = (val1 + val2) // 2
+        if ret % 2 == 1:
+            ret += 1
+        return ret
+
+    def auto_decrease_batch_size(self) -> int:
+        """Decrease batch size if default batch size isn't fit to current GPU device.
+
+        Returns:
+            int: Proper batch size possibly decreased as default value isn't fit
+        """
+        available_bs = 0
+        current_bs = self._default_bs
+        lowest_unavailable_bs = self._default_bs + 2
+
+        while True:
+            cuda_oom, max_memory_allocated = self._try_batch_size(current_bs)
+
+            # If GPU memory usage is too close to limit, CUDA OOM can be raised during training
+            if cuda_oom or max_memory_allocated > self._mem_upper_bound:
+                if current_bs < lowest_unavailable_bs:
+                    lowest_unavailable_bs = current_bs
+                current_bs = self._get_even_center_val(current_bs, available_bs)
+            else:
+                available_bs = current_bs
+                current_bs = self._get_even_center_val(current_bs, lowest_unavailable_bs)
+
+            if lowest_unavailable_bs - available_bs <= 2:
+                break
+
+        if available_bs == 0:
+            raise RuntimeError("Current device can't train model even with 2.")
+
+        return available_bs
+
+    def find_big_enough_batch_size(self, drop_last: bool = False) -> int:
+        """Find a big enough batch size.
+
+        This function finds a big enough batch size by training with various batch sizes.
+        It estimate a batch size using equation is estimated using training history.
+        The reason why using the word "big enough" is that it tries to find not maxmium but big enough value which uses
+        GPU memory between lower and upper bound.
+
+        Args:
+            drop_last (bool): Whether to drop the last incomplete batch.
+
+        Raises:
+            RuntimeError: If training with batch size 2 can't be run, raise an error.
+
+        Returns:
+            int: Big enough batch size.
+        """
+        estimated_bs = self._default_bs
+
+        # try default batch size
+        cuda_oom, bs_mem_usage = self._try_batch_size(estimated_bs)
+        if cuda_oom or bs_mem_usage > self._mem_upper_bound:
+            self._default_bs -= 2
+            if self._default_bs <= 0:
+                raise RuntimeError("Current device can't train model even with 2.")
+
+            return self.auto_decrease_batch_size()
+
+        # try default batch size + 2
+        estimated_bs += 2
+        if estimated_bs > self._max_bs:
+            return self._default_bs
+        cuda_oom, bs_mem_usage = self._try_batch_size(estimated_bs)
+        if cuda_oom or bs_mem_usage > self._mem_upper_bound:
+            return self._default_bs
+
+        # estimate batch size using equation
+        estimation_pct = 0.82
+        while True:
```

Comment:
This is the same as the comment above.

Reply:
As you know, it exits the loop if criteria is meet or estimated batch size is already tried. So I think it can always exit the loop.